README.rst
Eve-Swagger |latest-version|
============================

|build-status| |python-support|

Swagger_ extension for Eve_ powered RESTful APIs.

Usage
-----
.. code-block:: python

    from eve import Eve
    from eve_swagger import swagger, add_documentation

    app = Eve()
    app.register_blueprint(swagger)

    # required. See http://swagger.io/specification/#infoObject for details.
    app.config['SWAGGER_INFO'] = {
        'title': 'My Supercool API',
        'version': '1.0',
        'description': 'an API description',
        'termsOfService': 'my terms of service',
        'contact': {
            'name': 'nicola',
            'url': 'http://nicolaiarocci.com'
        },
        'license': {
            'name': 'BSD',
            'url': 'https://github.com/pyeve/eve-swagger/blob/master/LICENSE',
        },
        'schemes': ['http', 'https'],
    }

    # optional. Will use flask.request.host if missing.
    app.config['SWAGGER_HOST'] = 'myhost.com'

    # optional. Add/Update elements in the documentation at run-time without deleting subtrees.
    add_documentation({'paths': {'/status': {'get': {'parameters': [
        {
            'in': 'query',
            'name': 'foobar',
            'required': False,
            'description': 'special query parameter',
            'type': 'string'
        }]
    }}}})

    if __name__ == '__main__':
        app.run()

When the API is up and running, visit the ``/api-docs`` endpoint. The resulting
JSON can then be used with swagger tooling, like the Swagger UI or Swagger Editor:

.. image:: resources/swagger_editor.png

If you get the error "*Can't read from server. It may not have the appropriate
access-control-origin settings*" from Swagger UI, you might want to enable CORS
support with the ``X_DOMAINS`` and ``X_HEADERS`` configuration in your Eve
``settings.py``:

.. code-block:: python

    X_DOMAINS = ['http://localhost:8000',  # The domain where Swagger UI is running
                 'http://editor.swagger.io',
                 'http://petstore.swagger.io']
    X_HEADERS = ['Content-Type', 'If-Match']  # Needed for the "Try it out" buttons

For more information check the CORS documentation of `Swagger UI`_ and `Swagger
Editor`_.

Installation
------------
.. code-block::

    $ pip install eve-swagger


Description fields on the swagger docs
--------------------------------------

If you would like to include description fields to your swagger docs you can
include a description field in your schema validations in your ``settings.py``.
This can be done per field as well as on the resource-level.

As an example:

.. code-block:: python

    ...
    'description': 'Description of the user resource',
    'schema': {
        'userName': {
            'description': 'The username of the logged in user.',
            'type': 'string',
            'minlength': 1,
            'maxlength': 256,
            'required': True
        },
    }
    ...

**NOTE**: If you do use that feature make sure that the ``TRANSPARENT_SCHEMA_RULES``
in your ``settings.py`` is also turned ON, otherwise you will get complains from the
Cerberus library about "unknown field 'description' for field [yourFieldName]", or use 
a custom schema validator, as below:

.. code-block:: python

    from eve.io.mongo import Validator

    class MyValidator(Validator):
        def _validate_description(self, description, field, value):
            """ {'type': 'string'} """
            # Accept description attribute, used for swagger doc generation
            pass

    app = Eve(validator=MyValidator)


Disabling the documentation of a resource
-----------------------------------------

You can disable the documentation of a specific resource by adding a ``disable_documentation`` field
to the resource definition in ``settings.py``. This means that the resource will not show up in
the ``paths`` or ``definitions`` sections of the swagger docs.

.. code-block:: python

    ...
    'person': {
        'item_title': 'person',
        'disable_documentation': True,
        'schema': {...}
    }
    ...

Enabling the documentation of Eve event hooks
---------------------------------------------

By setting ``app.config['ENABLE_HOOK_DESCRIPTION']`` to ``True`` you can enable the description of all Eve event hooks.
This is done by showing the docstrings of the callback functions in the swagger docs under the appropriate ``paths``.

.. code-block:: python

    def foo(request, lookup):
        """ Do something before GETting all the people """
        pass
    def bar(response):
        """ Do something when you've fetched the database entries """
        pass
    ...
    app.config['ENABLE_HOOK_DESCRIPTION'] = True
    ...
    app.on_pre_GET_people += foo
    app.on_fetched_resource_people += bar

The swagger docs will now look like this:

.. code-block:: python

    "paths": {
        "/people": {
            "get": {
                ...,
                "description": "**Hooks**:\n* `on_pre_GET_people`:\n\n  * `foo`:\n\n    Do something before GETting all the people\n\n\n* `on_fetched_resource_people`:\n\n  * `bar`:\n\n    Do something when you've fetched the database entries\n\n"
            }
        }
    }

Which will be rendered by Swagger like this:

.. image:: resources/hook_description.png

Example fields on the docs
--------------------------

Like a description, an example can be added to a field.

.. code-block:: python

    ...
    'schema': {
        'lastName': {
            'example': 'Doe',
            'type': 'string',
            'minlength': 1,
        },
    }
    ...

The example is shown in the swagger ui in the model and the responses.

.. image:: resources/example_field.png

**NOTE**: As with the description, the field ``TRANSPARENT_SCHEMA_RULES``
must be enabled in your ``settings.py``, otherwise the Cerberus library 
will display an error about "unknown field 'example' for field[yourFieldName]".

If you do not want to use this rule for the entire schema, you can also use 
your own validator.

.. code-block:: python

    from eve.io.mongo import Validator

    class MyValidator(Validator):
        def _validate_example(self, example, field, value):
            if example and not isinstance(value, str):
                self._error(field, "Value must be a string")

    ...

    app = Eve(validator=MyValidator)

Copyright
---------
Eve-Swagger is an open source project by `Nicola Iarocci`_.
See the original LICENSE_ for more information.

.. |latest-version| image:: https://img.shields.io/pypi/v/eve-swagger.svg
   :alt: Latest version on PyPI
   :target: https://pypi.python.org/pypi/eve-swagger
.. |build-status| image:: https://travis-ci.org/pyeve/eve-swagger.svg?branch=master
   :alt: Build status
   :target: https://travis-ci.org/pyeve/eve-swagger
.. |python-support| image:: https://img.shields.io/pypi/pyversions/eve-swagger.svg
   :target: https://pypi.python.org/pypi/eve-swagger
   :alt: Python versions
.. |license| image:: https://img.shields.io/pypi/l/eve-swagger.svg
   :alt: Software license
   :target: https://github.com/pyeve/eve-swagger/blob/master/LICENSE

.. _Swagger: http://swagger.io/
.. _Eve: http://python-eve.org/
.. _`popular request`: https://github.com/pyeve/eve/issues/574
.. _LICENSE: https://github.com/pyeve/eve-swagger/blob/master/LICENSE
.. _`Nicola Iarocci`: http://nicolaiarocci.com
.. _`Swagger UI`: https://github.com/swagger-api/swagger-ui#enabling-cors
.. _`Swagger Editor`: https://github.com/swagger-api/swagger-editor/blob/master/docs/cors.md


CHANGES
Changelog
=========

Here you can see the full list of changes between each Eve-Swagger release.

In Development
--------------

- Add support for `nullable` property. (hermannsblumm)
- Fix: tests fail on flask_pymongo import. (`#63`_)

.. _`#63`: https://github.com/pyeve/eve-swagger/issues/63


Released
--------

Version 0.0.10
~~~~~~~~~~~~~~

Released on April 10, 2018.

- New: Add support for adding schema examples. Closes #59 (Nils Verheyen).

Version 0.0.9
~~~~~~~~~~~~~

Released on 6 March, 2018

- Fix: Override schemes in SWAGGER_INFO does not work (Cuong Manh Le).

Version 0.0.8
~~~~~~~~~~~~~

Released on 2 October, 2017

- Fix: `If-Match` header required when concurrency control is disabled. Closes
  #49 (Aleksi Pekkala).
- Fix: If the user provides a 'schemes' configuration value use that instead of
  auto-discovering the value. This is useful in the scenario where Eve is
  running on http behind a proxy running https, so auto-discovery would yield
  the wrong value (Jim Kennedy).
- New: add support for geometric data types `point`, `linestring`,
  `multilinestring', `polygon` and `multipolygon`. Closes #44 (Damien Aboss).

Version 0.0.7
~~~~~~~~~~~~~

Released on 13 March, 2017

- Fix: All responses claim to return a status code of 200 (Alexander
  Dietmüller).
- Fix: tests failure since jsonschema dropped Python 2.6. Closes #40.

Version 0.0.6
~~~~~~~~~~~~~

Released on 3 January, 2017

- New: support for swagger 2.0/openAPI 'readOnly' and 'pattern' rules
  (hermannsblum).
- Fix: crash if no docstring is available for hook (hermannsblum).
- Fix: support for data_relation nested in lists (Patrick Decat).
- Fix: handle missing fields and settings. Closes #31 (otibsa).

Version 0.0.5
~~~~~~~~~~~~~

Released on 25 October, 2016

- Fix: float type are not p:Groperly defined. Closes #13.
- Fix: README: fix comma that causes a validation error in example script (Luis
  Fernando Gomes).

- New: Ability to generate Swagger tags (Stratos Gerakakis). 
- New: add CORS support (otibsa).
- New: Python 2.6 compatibility (otibsa).
- New: Add a proper test suite. Closes #8 (otibsa).
- New: Move the path parameters (`/people/{personId}`) to their own section of
  the swagger doc. That cleans up the parameters entries in the paths sections
  by referencing the parameters (otibsa).
- New: Option to enable event hooks description to swagger documentation
  (otibsa).
- New: Support for data relations (otibsa).
- New: Option to disable resource documentation via `disable_documentation` in
  `settings.py` (otibsa).
- New: Ability to include description fields. Just add a `description` field in
  the schema definitions in `settings.py` (Stratos Gerakakis). 
- New: Add support for more eve features: `allowed`, `default`, `minlength`,
  `maxlength`, `min`, `max` (otibsa).
- New: Enable additional documentation to be injected at runtime (otibsa).
- New: Add tox and build server configuration (Peter Bittner). Addresses #8.
- New: Support for endpoint fields (swagger parameters). WIP. Closes #5.

Version 0.0.4
~~~~~~~~~~~~~

Released on 12 June, 2016

- New: Python 3 compatibility. Closes #6 (Naoko Reeves).

Version 0.0.3
~~~~~~~~~~~~~

Released on 7 June, 2016

- Fix: Crash on pip install. Closes #4.

Version 0.0.2
~~~~~~~~~~~~~

Released on 6 June, 2016

- Fix: AttributeError: 'module' object has no attribute 'name' when trying to
  register the Blueprint. Closes #3.

Version 0.0.1
~~~~~~~~~~~~~

Released on 4 June, 2016


eve_swagger/tests/tests.py
import json
from jsonschema import validate, ValidationError


if __name__ == '__main__':
    import sys
    import os
    import unittest
    sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__),
                                                 '..', '..')))

from eve_swagger.tests import TestBase  # noqa: E402


class TestEveSwagger(TestBase):
    def test_api(self):
        self.app.debug = True
        r = self.test_client.get('/api-docs')
        self.assertEqual(r.status_code, 200)

    def test_json(self):
        self.app.debug = True
        r = self.test_client.get('/api-docs')
        s = r.get_data().decode('utf-8')  # raises UnicodeError
        json.loads(s)  # raises ValueError

    def test_doc_is_dict(self):
        doc = self.swagger_doc
        self.assertIsInstance(doc, dict)

    def test_info(self):
        doc = self.swagger_doc

        self.assertIn('info', doc)
        self.assertIn('title', doc['info'])
        self.assertTrue(isinstance(doc['info']['title'], u''.__class__))
        self.assertIn('version', doc['info'])
        self.assertTrue(isinstance(doc['info']['version'], u''.__class__))

    def test_paths(self):
        doc = self.swagger_doc
        url = self.domain['people']['url']
        item_title = self.domain['people']['item_title']

        self.assertIn('paths', doc)
        self.assertIsInstance(doc['paths'], dict)
        self.assertIn('/'+url, doc['paths'])
        self.assertIn('/%s/{%sId}' % (url, item_title.lower()), doc['paths'])

    def test_definitions(self):
        doc = self.swagger_doc
        item_title = self.domain['people']['item_title']

        self.assertIn('definitions', doc)
        self.assertIsInstance(doc['definitions'], dict)
        self.assertIn(item_title, doc['definitions'])
        self.assertIn('properties', doc['definitions'][item_title])
        self.assertEqual(
            set(doc['definitions'][item_title]['properties'].keys()),
            set(['name', 'job', 'email', 'position', '_id',
                 'relations', 'location']))

    def test_definitions_are_jsonschema(self):
        doc = self.swagger_doc
        item_title = self.domain['people']['item_title']

        # validate if schema is jsonschema
        try:
            validate({}, doc['definitions'][item_title])
        except ValidationError:
            # validation errors are only thrown after schema is checked
            self.assertTrue(True)

    def test_parameters_people(self):
        doc = self.swagger_doc
        item_title = self.domain['people']['item_title']
        lookup_field = self.domain['people']['item_lookup_field']

        self.assertIn('parameters', doc)
        self.assertIn(item_title+'_'+lookup_field, doc['parameters'])

    def test_resource_description(self):
        doc = self.swagger_doc
        item_title = self.domain['people']['item_title']

        self.assertEqual('the people resource',
                         doc['definitions'][item_title]['description'])

    def test_field_description(self):
        doc = self.swagger_doc
        item_title = self.domain['people']['item_title']
        props = doc['definitions'][item_title]['properties']

        self.assertEqual('the last name of the person',
                         props['name']['description'])

    def test_field_example(self):
        doc = self.swagger_doc
        item_title = self.domain['people']['item_title']
        props = doc['definitions'][item_title]['properties']

        self.assertEqual('Doe', props['name']['example'])

    def test_disabled_resource(self):
        doc = self.swagger_doc
        url = self.domain['disabled_resource']['url']
        item_title = self.domain['disabled_resource']['item_title']
        lookup_field = self.domain['disabled_resource']['item_lookup_field']

        self.assertNotIn('/'+url, doc['paths'])
        self.assertNotIn('/%s/{%sId}' % (url, item_title.lower()),
                         doc['paths'])
        self.assertNotIn(item_title, doc['definitions'])
        self.assertNotIn(item_title+'_'+lookup_field, doc['parameters'])

    def test_data_relation_source_field(self):
        doc = self.swagger_doc

        source_field = self.domain['people']['item_title']+'_job'
        self.assertIn(source_field, doc['definitions'])

    def test_reference_to_data_relation_source_field(self):
        doc = self.swagger_doc
        people_it = self.domain['people']['item_title']
        people_props = doc['definitions'][people_it]['properties']
        dr_1_it = self.domain['dr_resource_1']['item_title']
        dr_1_props = doc['definitions'][dr_1_it]['properties']
        key = people_it + '_job'

        self.assertIn('$ref', people_props['job'])
        self.assertEqual('#/definitions/%s' % key,
                         people_props['job']['$ref'])
        self.assertIn(key, doc['definitions'])

        self.assertIn('$ref', dr_1_props['copied_field_with_description'])
        self.assertEqual('#/definitions/%s' % key,
                         dr_1_props['copied_field_with_description']['$ref'])

        key = people_it + '__id'
        people_rels_props = people_props['relations']['items']['properties']
        self.assertIn('$ref', people_rels_props['relation'])
        self.assertEqual('#/definitions/%s' % key,
                         people_rels_props['relation']['$ref'])
        self.assertIn(key, doc['definitions'])

    def test_data_relation_extended_description(self):
        doc = self.swagger_doc
        item_title = self.domain['dr_resource_1']['item_title']
        lookup_field = self.domain['dr_resource_1']['item_lookup_field']
        par = doc['parameters'][item_title+'_'+lookup_field]
        people_it = self.domain['people']['item_title']

        self.assertIn('description', par)
        self.assertEqual(
            'foobar copied_field (links to {0}_job)'.format(people_it),
            par['description'])

    def test_data_relation_copied_description(self):
        doc = self.swagger_doc
        item_title = self.domain['dr_resource_2']['item_title']
        lookup_field = self.domain['dr_resource_2']['item_lookup_field']
        par = doc['parameters'][item_title+'_'+lookup_field]
        people_it = self.domain['people']['item_title']

        self.assertIn('description', par)
        self.assertEqual(
            'the job of the person (links to {0}_job)'.format(people_it),
            par['description'])

    def test_header_parameters(self):
        doc = self.swagger_doc
        url = self.domain['people']['url']
        item_title = self.domain['people']['item_title']
        url = '/%s/{%sId}' % (url, item_title.lower())

        header_parameters = []
        # assume that header parameters are equal for PUT, PATCH, and DELETE
        for method in ['put', 'patch', 'delete']:
            for p in doc['paths'][url][method]['parameters']:
                if 'in' not in p:
                    continue
                if p['in'] == 'header':
                    if method in ['patch', 'delete']:
                        # already added in 'put'
                        self.assertIn(p, header_parameters)
                    else:
                        header_parameters += [p]

        self.assertTrue(len(header_parameters) == 1)
        h = header_parameters[0]
        self.assertIn('name', h)
        self.assertEqual(h['name'], 'If-Match')
        self.assertTrue(h['required'])

    def test_header_parameters_without_concurrency_control(self):
        url = self.domain['people']['url']
        item_title = self.domain['people']['item_title']
        url = '/%s/{%sId}' % (url, item_title.lower())

        def get_etag_param(doc):
            params = doc['paths'][url]['delete']['parameters']
            return next(p for p in params if p.get('name') == 'If-Match')

        self.app.config['IF_MATCH'] = False
        etag_param = get_etag_param(self.get_swagger_doc())
        self.assertFalse(etag_param['required'])

        self.app.config['IF_MATCH'] = True
        self.app.config['ENFORCE_IF_MATCH'] = False
        etag_param = get_etag_param(self.get_swagger_doc())
        self.assertFalse(etag_param['required'])

    def test_cors_without_origin(self):
        self.app.config['X_DOMAINS'] = ['http://example.com']
        self.app.config['X_HEADERS'] = ['Origin', 'X-Requested-With',
                                        'Content-Type', 'Accept']
        self.app.config['X_MAX_AGE'] = 2000
        r = self.test_client.get('/api-docs')

        self.assertEqual(r.status_code, 200)
        self.assertNotIn('Access-Control-Allow-Origin', r.headers)
        self.assertNotIn('Access-Control-Allow-Headers', r.headers)
        self.assertNotIn('Access-Control-Allow-Methods', r.headers)
        self.assertNotIn('Access-Control-Max-Age', r.headers)
        self.assertNotIn('Access-Control-Expose-Headers', r.headers)

    def test_cors_with_origin(self):
        self.app.config['X_DOMAINS'] = 'http://example.com'
        self.app.config['X_HEADERS'] = ['Origin', 'X-Requested-With',
                                        'Content-Type', 'Accept']
        self.app.config['X_MAX_AGE'] = 2000
        r = self.test_client.get('/api-docs',
                                 headers={'Origin': 'http://example.com'})

        self.assertEqual(r.status_code, 200)
        self.assertIn('Access-Control-Allow-Origin', r.headers)
        self.assertEqual(r.headers['Access-Control-Allow-Origin'],
                         'http://example.com')
        self.assertIn('Access-Control-Allow-Headers', r.headers)
        self.assertEqual(
            set(r.headers['Access-Control-Allow-Headers'].split(', ')),
            set(['Origin', 'X-Requested-With', 'Content-Type', 'Accept']))
        self.assertIn('Access-Control-Allow-Methods', r.headers)
        self.assertEqual(
            set(r.headers['Access-Control-Allow-Methods'].split(', ')),
            set(['HEAD', 'OPTIONS', 'GET']))
        self.assertIn('Access-Control-Max-Age', r.headers)
        self.assertEqual(r.headers['Access-Control-Max-Age'],
                         '2000')

    def test_tags(self):
        doc = self.swagger_doc
        item_title = self.domain['people']['item_title']

        self.assertIn('tags', doc)
        tag_names = [tag['name'] for tag in doc['tags']]
        self.assertTrue(len([n for n in tag_names if n == item_title]) == 1)

        url = '/' + self.domain['people']['url']
        for m in ['get', 'post', 'delete']:
            path_m = doc['paths'][url][m]
            self.assertIn('tags', path_m)
            self.assertIn(item_title, path_m['tags'])

        url = '/%s/{%sId}' % (self.domain['people']['url'], item_title.lower())
        for m in ['get', 'patch', 'put', 'delete']:
            path_m = doc['paths'][url][m]
            self.assertIn('tags', path_m)
            self.assertIn(item_title, path_m['tags'])

    def test_status_codes(self):
        doc = self.swagger_doc
        url = self.domain['people']['url']

        people = doc['paths']['/'+url]
        self.assertIn('200', people['get']['responses'])
        self.assertIn('201', people['post']['responses'])
        self.assertIn('204', people['delete']['responses'])

        item_title = self.domain['people']['item_title']
        person = doc['paths']['/%s/{%sId}' % (url, item_title.lower())]
        self.assertIn('200', person['get']['responses'])
        self.assertIn('200', person['patch']['responses'])
        self.assertIn('200', person['put']['responses'])
        self.assertIn('204', person['delete']['responses'])

    def test_schemes_override(self):
        self.app.config['SWAGGER_INFO']['schemes'] = ['https']
        r = self.test_client.get('/api-docs')
        self.assertEqual(r.status_code, 200)
        s = r.get_data().decode('utf-8')
        d = json.loads(s)
        self.assertEqual(d.get('schemes'), ['https'])


if __name__ == '__main__':
    unittest.main()


eve_swagger/tests/__init__.py
import sys
import os
import json

import eve
import eve_swagger
from pymongo import MongoClient
from eve_swagger.tests.test_settings import MONGO_HOST, MONGO_PORT, \
    MONGO_USERNAME, MONGO_PASSWORD, MONGO_DBNAME

if sys.version_info >= (2, 7):
    import unittest  # noqa
else:
    import unittest2 as unittest  # noqa


class TestBase(unittest.TestCase):
    def setUp(self, settings=None):
        self.this_directory = os.path.dirname(os.path.realpath(__file__))
        if settings is None:
            settings = os.path.join(self.this_directory, 'test_settings.py')

        self.setupDB()

        self.settings = settings
        self.app = eve.Eve(settings=self.settings)

        self.app.register_blueprint(eve_swagger.swagger)
        self.app.config['SWAGGER_INFO'] = {
            'title': 'Test eve-swagger',
            'version': '0.0.1'
        }

        self.test_client = self.app.test_client()
        self.domain = self.app.config['DOMAIN']

        self.swagger_doc = self.get_swagger_doc()

    def tearDown(self):
        del self.app
        self.dropDB()

    def setupDB(self):
        self.connection = MongoClient(MONGO_HOST, MONGO_PORT)
        self.connection.drop_database(MONGO_DBNAME)
        if MONGO_USERNAME:
            self.connection[MONGO_DBNAME].add_user(MONGO_USERNAME,
                                                   MONGO_PASSWORD)

    def dropDB(self):
        self.connection = MongoClient(MONGO_HOST, MONGO_PORT)
        self.connection.drop_database(MONGO_DBNAME)
        self.connection.close()

    def get_swagger_doc(self):
        r = self.test_client.get('/api-docs')
        return self.parse_response(r)

    def parse_response(self, r):
        try:
            v = json.loads(r.get_data().decode('utf-8'))
        except ValueError:
            v = None
        return v


eve_swagger/validation.py
# -*- coding: utf-8 -*-
"""
    eve-swagger.validation
    ~~~~~~~~~~~~~~~~~~~~~~
    swagger.io extension for Eve-powered REST APIs.

    :copyright: (c) 2015 by Nicola Iarocci.
    :license: BSD, see LICENSE for more details.
"""
from eve.exceptions import ConfigException
from flask import current_app as app
from cerberus import Validator

import eve_swagger

try:
    from urllib.parse import urlparse
except ImportError:
    from urlparse import urlparse


def validate_info():
    v = Validator()
    schema = {
        'title': {'required': True, 'type': 'string'},
        'version': {'required': True, 'type': 'string'},
        'description': {'type': 'string'},
        'termsOfService': {'type': 'string'},
        'contact': {
            'type': 'dict',
            'schema': {
                'name': {'type': 'string'},
                'url': {'type': 'string', 'validator': _validate_url},
                'email': {
                    'type': 'string',
                    'regex':
                    '^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$'
                }
            }
        },
        'license': {
            'type': 'dict',
            'schema': {
                'name': {'type': 'string', 'required': True},
                'url': {'type': 'string', 'validator': _validate_url}
            }
        },
        'schemes': {
            'type': 'list',
            'schema': {
                'type': 'string'
            }
        }
    }
    if eve_swagger.INFO not in app.config:
        raise ConfigException('%s setting is required in Eve configuration.' %
                              eve_swagger.INFO)

    if not v.validate(app.config[eve_swagger.INFO], schema):
        raise ConfigException('%s is misconfigured: %s' % (
            eve_swagger.INFO, v.errors))


def _validate_url(field, value, error):
    # TODO probably too weak
    o = urlparse(value)
    if not bool(o.scheme):
        error(field, 'Invalid URL')


eve_swagger/paths.py
# -*- coding: utf-8 -*-
"""
    eve-swagger.paths
    ~~~~~~~~~~~~~~~~~
    swagger.io extension for Eve-powered REST APIs.

    :copyright: (c) 2015 by Nicola Iarocci.
    :license: BSD, see LICENSE for more details.
"""
from textwrap import dedent

from flask import current_app as app

from eve_swagger import OrderedDict


# TODO consider adding at least a 'schema' property to response objects
# TODO take auth into consideration


def paths():
    paths = OrderedDict()
    for resource, rd in app.config['DOMAIN'].items():
        if (rd.get('disable_documentation')
                or resource.endswith('_versions')):
            continue

        methods = rd['resource_methods']
        if methods:
            url = '/%s' % rd['url']
            paths[url] = _resource(resource, rd, methods)

        methods = rd['item_methods']
        if methods:
            item_id = '%sId' % rd['item_title'].lower()
            url = '/%s/{%s}' % (rd['url'], item_id)
            paths[url] = _item(resource, rd, methods)

    return paths


def _resource(resource, rd, methods):
    item = OrderedDict()
    describe_hooks = app.config.get('ENABLE_HOOK_DESCRIPTION', False)
    if 'GET' in methods:
        item['get'] = get_response(rd)
    if 'POST' in methods:
        item['post'] = post_response(rd)
    if 'DELETE' in methods:
        item['delete'] = delete_response(rd)

    if describe_hooks:
        for m in methods:
            hook_desc = _hook_descriptions(resource, m)
            if hook_desc != '':
                item[m.lower()]['description'] = '**Hooks**:' + hook_desc

    return item


def _item(resource, rd, methods):
    item = OrderedDict()
    describe_hooks = app.config.get('ENABLE_HOOK_DESCRIPTION', False)
    if 'GET' in methods:
        item['get'] = getitem_response(rd)
    if 'PUT' in methods:
        item['put'] = put_response(rd)
    if 'PATCH' in methods:
        item['patch'] = patch_response(rd)
    if 'DELETE' in methods:
        item['delete'] = deleteitem_response(rd)

    if describe_hooks:
        for m in methods:
            hook_desc = _hook_descriptions(resource, m, item=True)
            if hook_desc != '':
                item[m.lower()]['description'] = '**Hooks**:' + hook_desc

    return item


def get_ref_schema(rd):
    return {'$ref': '#/definitions/%s' % rd['item_title']}


def get_parameters(rd):
    return OrderedDict([
        ('in', 'body'),
        ('name', rd['item_title']),
        ('required', True),
        ('schema', get_ref_schema(rd)),
    ])


def get_response(rd):
    title = rd['resource_title']
    return OrderedDict([
        ('summary', 'Retrieves one or more %s' % title),
        ('responses', {'200': {
            'description': 'An array of %s' % title,
            'schema': {
                'type': 'array',
                'items': get_ref_schema(rd)}}}),
        ('tags', [rd['item_title']])
    ])


def post_response(rd):
    return OrderedDict([
        ('summary', 'Stores one or more %s' % rd['resource_title']),
        ('parameters', [get_parameters(rd)]),
        ('responses', {'201':
                       {'description': 'operation has been successful'}}),
        ('tags', [rd['item_title']])
    ])


def delete_response(rd):
    return OrderedDict([
        ('summary', 'Deletes all %s' % rd['resource_title']),
        ('responses', {'204':
                       {'description': 'operation has been successful'}}),
        ('tags', [rd['item_title']])
    ])


def getitem_response(rd):
    title = rd['item_title']
    return OrderedDict([
        ('summary', 'Retrieves a %s document' % title),
        ('responses', {
            '200': {
                'description': '%s document fetched successfully' % title,
                'schema': get_ref_schema(rd)
            },

        }),
        ('parameters', [id_parameter(rd)]),
        ('tags', [rd['item_title']])
    ])


def put_response(rd):
    title = rd['item_title']
    return OrderedDict([
        ('summary', 'Replaces a %s document' % title),
        ('responses', {
            '200': {
                'description': '%s document replaced successfully' % title
            }
        }),
        ('parameters', [id_parameter(rd),
                        get_parameters(rd),
                        header_parameters()]),
        ('tags', [rd['item_title']])
    ])


def patch_response(rd):
    title = rd['item_title']
    return OrderedDict([
        ('summary', 'Updates a %s document' % title),
        ('responses', {
            '200': {
                'description': '%s document updated successfully' % title
            }
        }),
        ('parameters', [id_parameter(rd),
                        get_parameters(rd),
                        header_parameters()]),
        ('tags', [rd['item_title']])
    ])


def deleteitem_response(rd):
    title = rd['item_title']
    return OrderedDict([
        ('summary', 'Deletes a %s document' % title),
        ('responses', {
            '204': {
                'description': '%s document deleted successfully' % title
            }
        }),
        ('parameters', [id_parameter(rd), header_parameters()]),
        ('tags', [rd['item_title']])
    ])


def id_parameter(rd):
    return {'$ref': '#/parameters/{0}_{1}'.format(rd['item_title'],
                                                  rd['item_lookup_field'])}


def header_parameters():
    r = OrderedDict()
    r['in'] = 'header'
    r['name'] = 'If-Match'
    r['description'] = 'Current value of the _etag field'
    r['required'] = app.config['IF_MATCH'] and app.config['ENFORCE_IF_MATCH']
    r['type'] = 'string'
    return r


def _hook_descriptions(resource, method, item=False):
    if method == 'GET':
        if item is True:
            events = ['on_pre_GET',
                      'on_pre_GET_' + resource,
                      'on_fetched_item',
                      'on_fetched_item_' + resource,
                      'on_post_GET',
                      'on_post_GET_' + resource]
        else:
            events = ['on_pre_GET',
                      'on_pre_GET_' + resource,
                      'on_fetched_resource',
                      'on_fetched_resource_' + resource,
                      'on_post_GET',
                      'on_post_GET_' + resource]

    if method == 'POST':
        events = ['on_pre_POST',
                  'on_pre_POST_' + resource,
                  'on_insert',
                  'on_insert_' + resource,
                  'on_inserted',
                  'on_inserted_' + resource,
                  'on_post_POST',
                  'on_post_POST_' + resource]
    if method == 'PUT':
        events = ['on_pre_PUT',
                  'on_pre_PUT_' + resource,
                  'on_replace',
                  'on_replace_' + resource,
                  'on_replaced',
                  'on_replaced_' + resource,
                  'on_post_PUT',
                  'on_post_PUT_' + resource]
    if method == 'PATCH':
        events = ['on_pre_PATCH',
                  'on_pre_PATCH_' + resource,
                  'on_update',
                  'on_update_' + resource,
                  'on_updated',
                  'on_updated_' + resource,
                  'on_post_PATCH',
                  'on_post_PATCH_' + resource]
    if method == 'DELETE':
        if item is True:
            events = ['on_pre_DELETE',
                      'on_pre_DELETE_' + resource,
                      'on_delete_item',
                      'on_delete_item_' + resource,
                      'on_deleted_item',
                      'on_deleted_item_' + resource,
                      'on_post_DELETE',
                      'on_post_DELETE_' + resource]
        else:
            events = ['on_pre_DELETE',
                      'on_pre_DELETE_' + resource,
                      'on_delete_resource',
                      'on_delete_resource_' + resource,
                      'on_deleted_resource',
                      'on_deleted_resource_' + resource,
                      'on_post_DELETE',
                      'on_post_DELETE_' + resource]

    res = ''
    for e in events:
        callbacks = getattr(app, e)
        if len(callbacks) > 0:
            res += '\n* `' + e + '`:\n\n'
        for cb in callbacks:
            if cb.__doc__:
                s = '\n    '
                s += '\n    '.join(dedent(cb.__doc__).strip().split('\n'))
                res += '  * `' + cb.__name__ + '`:\n' + s + '\n\n'
            else:
                # there is no docstring provided, still add the hook name for
                # information
                res += '  * `' + cb.__name__ + '`:\nno documentation\n\n'
    return res
